### PR TITLE
parser.clj (parse-info): Return :type for fields and methods

### PR DESCRIPTION
**Pull request by abo-abo**
Allows e.g. to have the following sample of tags for
java.util.ArrayList.java:

    int MAX_ARRAY_SIZE
    void grow (int minCapacity)
    int hugeCapacity (int minCapacity)
    int size ()
    boolean isEmpty ()
    boolean contains (Object o)
    int indexOf (Object o)
    int lastIndexOf (Object o)
    Object clone ()
    Object[] toArray ()
    Object[] toArray (Object[] a)

### CODE REVIEW

File: src/cider/nrepl/middleware/util/java/parser.clj [hunk 1]
```diff
@@ -216,10 +216,12 @@
   MethodDoc
   (parse-info [m]
     {:argtypes (mapv #(-> ^Parameter % .type typesym) (.parameters m))
-     :argnames (mapv #(-> ^Parameter % .name symbol) (.parameters m))})
+     :argnames (mapv #(-> ^Parameter % .name symbol) (.parameters m))
+     :type (str (. m returnType))})
```
Comment (xiongtx) [inline src/cider/nrepl/middleware/util/java/parser.clj]: `(.returnType m)`

File: src/cider/nrepl/middleware/util/java/parser.clj [hunk 2]
```diff
@@ -216,10 +216,12 @@
...
 
   FieldDoc
-  (parse-info [f])
+  (parse-info [f]
+    {:type (str (. f type))})
```
Comment (xiongtx) [inline src/cider/nrepl/middleware/util/java/parser.clj]: `(.type f)`

File: src/cider/nrepl/middleware/util/java/parser.clj (hunk 1 above)
Comment (abo-abo) [reply src/cider/nrepl/middleware/util/java/parser.clj]: I prefer `(. m returnType)`. In line with Java being a language designed for an IDE, you never type out:

```java
m.returnType
```

You type `m.` and get a list of completions based on the class of `m`.

I do the same in Clojure: type `(. m )` and get a list of completions for the class of `m`.

Comment (bbatsov) [reply src/cider/nrepl/middleware/util/java/parser.clj]: But it's not like this in Clojure - as you don't really know the type of the receiver you get completions for possible classes anyways. At any rate - consistency in a codebase is important and we've always used the syntax suggested by @xiongtx so far, so please change this.

**Comment by bbatsov** [comment]
How do you plan to use this on the client side?

**Comment by abo-abo** [comment]
See commit message, to generate semantic tags.

**Comment by bbatsov** [comment]
Got it. I guess you meant to write "semantic", but wrote "sample" instead in the commit message.

**Comment by abo-abo** [comment]
> Got it. I guess you meant to write "semantic", but wrote "sample" instead in the commit message.

By "semantic" I mean tags that are retrieved from the AST, not just extracted from the source with elaborate regexes.

By "sample" I mean a sample of `imenu`-like strings for java.util.ArrayList.java; I don't include them all :)

**Comment by bbatsov** [comment]
Now I finally got it. :D Can you move this PR here https://github.com/clojure-emacs/orchard/ ?

I'm now in the process of decoupling the middleware from the generic code and this file is now living here https://github.com/clojure-emacs/orchard/blob/master/src/orchard/java/parser.clj

**Pull request closed without merging**